proposals/0000-surface-async-stream-errors-to-http.md

```diff
@@ -0,0 +1,224 @@
+- https://nodejs.org/api/http.html#http_event_clienterror
+- `http2session.destroy([error][, code])`
+- `http2session.goaway([code[, lastStreamID[, opaqueData]]])`
+- https://nodejs.org/api/http2.html#http2_destruction
+- `http2stream.close(code[, callback])`
+- `response.addTrailers(headers)`
+
+### HTTP ecosystem considerations
+
+It should be possible for developers to selectively turn off this error-signaling behavior on a per-request basis, in case of misbehaving proxies or known-bad clients.
+
+Research needed into how common reverse proxies or load balancers, such as nginx, handle error-signaled HTTP streams:
+
+- [Real-world examples of how that can happen, including the comment at the very bottom](https://rhodesmill.org/brandon/2013/chunked-wsgi/)
+- [Atmosphere long-polling on nginx – chunked encoding error](https://rijulaggarwal.wordpress.com/2018/01/10/atmosphere-long-polling-on-nginx-chunked-encoding-error/)
+
+In the common case of a backend speaking HTTP/1.1 to a reverse proxy/front-end terminator/CDN/etc. that translates to HTTP/2 for browsers, what can be done?
+
+## Migration strategy
+
+> Is this proposal backwards incompatible? Does this proposal replace an existing feature/pattern? If so, can we safely and automatically migrate existing apps/components to this proposal? How?
+
+> If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+
+## Drawbacks
+
+So far, Marko is totally agnostic as to what kind of stream it’s outputting to. You could use it to stream to a file, a UNIX pipe, or something really unexpected like an FTP stream transmission or a chat protocol.
+
+It’s possible the server runtime might expose a different method to avoid breaking such cases, like `template.streamToHttp()`.
+
+However, there’s already a lot about Marko’s server rendering that only makes sense delivered over HTTP. For example, if you were rendering to a file to be served later, you’d want to buffer the component init scripts and you’d never want `client-reorder` enabled.
+
+## Alternatives
+
+The impact of not doing this would be no change; we would continue exposing Marko authors to the risks mentioned in the Motivation section.
+
+In the past, browsers other than Internet Explorer accepted `multipart/x-mixed-replace` responses to be loaded in `target="_top"`, and that could potentially be used to swap to a new copy of the response-in-progress but with error-indicating headers, or even a replacement error page. However, modern browsers have quietly dropped support for `multipart/x-mixed-replace` in the top browsing context.
+
+It _might_ be possible to write `<meta http-equiv="refresh" content="0;url=${errorPageLocation}">`. It’s not _supposed_ to be outside the `<head>`, but I still like it better than an error redirect via inline `<script>`. ([some context on how search engines could better understand this is a temporary redirect, not a permanent one](https://twitter.com/JohnMu/status/969486943351394304))
```

Comment:
Note to self: make sure browsers don’t cache the HTML if we navigate away before the page finishes rendering